statefun-sdk-js/src/handler.ts

```diff
@@ -0,0 +1,247 @@
+
+async function tryHandle(toFunctionBytes: Buffer | Uint8Array, fns: any): Promise<Buffer | Uint8Array> {
+    //
+    // setup
+    //
+    const toFn = PB_ToFn.deserializeBinary(toFunctionBytes);
+    const pbInvocationBatchRequest = toFn.getInvocation();
+    if (pbInvocationBatchRequest === null) {
+        throw new Error("An empty invocation request");
+    }
+    const targetAddress = pbAddressToSdkAddress(pbInvocationBatchRequest.getTarget());
+    if (targetAddress === null) {
+        throw new Error("Missing target address");
+    }
+    const fnSpec = findTargetFunctionSpec(fns, targetAddress);
+    const {missing, values, storage} = AddressScopedStorageFactory.tryCreateAddressScopedStorage(pbInvocationBatchRequest, fnSpec.valueSpecs);
+    if (missing !== null) {
+        return respondImmediatelyWithMissingContext(missing);
+    }
+    if (values === undefined || values === null) {
+        throw new Error("Unexpected internal error. Could not create an address scoped storage.");
+    }
+    //
+    // apply the batch
+    //
+    const internalContext = new InternalContext()
+    const context = new Context(targetAddress, storage, internalContext);
+    await applyBatch(pbInvocationBatchRequest, context, internalContext, fnSpec.fn);
+    //
+    // collect the side effects
+    //
+    let pbInvocationResponse = new PB_InvocationResponse();
+    collectStateMutations(values, pbInvocationResponse);
+    collectOutgoingMessages(internalContext.sent, pbInvocationResponse);
+    collectEgress(internalContext.egress, pbInvocationResponse);
+    collectDelayedMessage(internalContext.delayed, pbInvocationResponse);
+
+    let fromFn = new PB_FromFn();
+    fromFn.setInvocationResult(pbInvocationResponse)
+    return fromFn.serializeBinary();
+}
+
+// noinspection JSValidateJSDoc
+/**
+ * @param {?proto.io.statefun.sdk.reqreply.ToFunction.InvocationBatchRequest} pbInvocationBatchRequest
+ * @param {Context} context user facing context
+ * @param {InternalContext} internalContext
+ * @param {*} fn the function to apply
+ */
+async function applyBatch(pbInvocationBatchRequest: any, context: Context, internalContext: InternalContext, fn: any) {
+    for (let invocation of pbInvocationBatchRequest.getInvocationsList()) {
+        internalContext.caller = pbAddressToSdkAddress(invocation.getCaller());
+        const message = new Message(context.self, invocation.getArgument());
+        const maybePromise = fn(context, message);
+        if (maybePromise instanceof Promise) {
+            await maybePromise;
+        }
+    }
+}
+
+// ----------------------------------------------------------------------------------------------------
+// Side Effect Collection
+// ----------------------------------------------------------------------------------------------------
+
+function collectStateMutations(values: Value[], pbInvocationResponse: any) {
```

Comment:
The comments from above apply to most other functions here as well, I'll stop repeating them :-)